File: standard-udfs-spark/src/test/scala/com/linkedin/stdudfs/spark/examples/TestMapKeySetFunctionWrapper.scala
package com.linkedin.stdudfs.spark.examples

import com.linkedin.stdudfs.spark.common.AssertSparkExpression._
import org.testng.annotations.{BeforeClass, Test}

import scala.collection.mutable.WrappedArray._

class TestMapKeySetFunctionWrapper {

  @BeforeClass
  def registerFunction(): Unit = {
    registerStandardUdf(
      "map_key_set",
      classOf[MapKeySetFunctionWrapper]
    )
  }

  @Test
  def testMapKeySet(): Unit = {
    assertFunction("map_key_set(map(1, 4, 2, 5, 3, 6))", make(Array(1, 2, 3)))
    assertFunction("map_key_set(map('1', '4', '2', '5', '3', '6'))", make(Array("1", "2", "3")))
    assertFunction("map_key_set(null)", null)
  }
}

File: standard-udfs-presto/src/test/java/com/linkedin/stdudfs/presto/examples/TestMapKeySetFunctionWrapper.java
package com.linkedin.stdudfs.presto.examples;

import com.facebook.presto.operator.scalar.AbstractTestFunctions;
import com.facebook.presto.spi.type.ArrayType;
import com.google.common.collect.ImmutableList;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.Test;

import static com.facebook.presto.spi.type.IntegerType.*;
import static com.facebook.presto.spi.type.VarcharType.*;
import static com.facebook.presto.type.UnknownType.*;


public class TestMapKeySetFunctionWrapper extends AbstractTestFunctions {

  @BeforeClass
  public void registerFunction() {
    registerScalarFunction(new MapKeySetFunctionWrapper());
  }

  @Test
  public void testMapKeySet() {

    assertFunction("map_key_set(map(array[1, 2, 3], array[4, 5, 6]))", new ArrayType(INTEGER),
        ImmutableList.of(1, 2, 3));

    assertFunction("cast (map_key_set(map(array['1', '2', '3'], array['4', '5', '6'])) as array(varchar))",
        new ArrayType(VARCHAR), ImmutableList.of("1", "2", "3"));

    assertFunction("map_key_set(null)", new ArrayType(UNKNOWN), null);
  }
}
